### PR TITLE
Avoid Calcite classes in SQL api

### PR DESCRIPTION
**Pull request by TomaszGaweda**
Avoid Calcite's RexNode in the SqlConnector API.

This will enable us to write SqlConnectors outside of the hazelcast-sql module. Without it, we cannot really implement the interface, as other modules won't have an access to shaded Calcite classes

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

**Comment by TomaszGaweda** [comment]
run-lab-run

**Comment by TomaszGaweda** [comment]
run-mongo-tests

**Comment by TomaszGaweda** [comment]
Ups, wrong PR for the trigger :D

**Comment by TomaszGaweda** [comment]
run-ee-tests

**Comment by TomaszGaweda** [comment]
run-nightly-tests

**Comment by TomaszGaweda** [comment]
I build https://github.com/hazelcast/hazelcast/pull/23508 with this PR and moved MongoDB SQL code to mongodb module.

Ran mongodb:
`docker run --rm --name mongo -p 27017:27017 -d mongo:latest`
and 
```
db.createCollection('test')
db.test.insert({id: "Test", age: 1337})
```

Started `hz-start` from distribution zip and ran:
```
create mapping test (
    id varchar
)
    TYPE MongoDB
options (
    'connectionString' = 'mongodb://localhost:27017',
    'database' = 'test',
    'collection' = 'test'
);


select * from test;
```

Query worked :)

**Comment by TomaszGaweda** [comment]
Required shade plugin config in MongoDB SQL module:
```xml

            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-shade-plugin</artifactId>
                <version>${maven.shade.plugin.version}</version>
                <executions>
                    <execution>
                        <id>fat-shaded-jar</id>
                        <phase>package</phase>
                        <goals>
                            <goal>shade</goal>
                        </goals>
                        <configuration>
                            <filters>
                                <filter>
                                    <artifact>*:*</artifact>
                                    <excludes>
                                        <exclude>module-info.class</exclude>
                                        <exclude>META-INF/*.SF</exclude>
                                        <exclude>META-INF/*.DSA</exclude>
                                        <exclude>META-INF/*.RSA</exclude>
                                    </excludes>
                                </filter>
                            </filters>
                            <relocations>
                                <relocation>
                                    <pattern>org.apache.calcite</pattern>
                                    <shadedPattern>${relocation.root}.org.apache.calcite</shadedPattern>
                                </relocation>
                            </relocations>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

**Comment by TomaszGaweda** [comment]
run-nightly-tests

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    public static @Nonnull CalciteNode filter(RexNode node) {
+        return new CalciteNode(node);
+    }
+
+    public static @Nonnull List<CalciteNode> projection(@Nonnull List<RexNode> nodes) {
+        return nodes.stream()
+                .map(CalciteNode::new)
+                .collect(toList());
+    }
```
Comment (frant-hartm) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java:59]: These method names are wrong. Should be something like create/adapt/convert or similar.
There is no filtering nor projecting.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java [hunk 2]
```diff
@@ -228,8 +229,8 @@ public Vertex onFullScan(FullScanPhysicalRel rel) {
         dagBuildContext.setRel(rel);
         return getJetSqlConnector(table).fullScanReader(
                 dagBuildContext,
-                rel.filter(),
-                rel.projection(),
+                CalciteNode.filter(rel.filter()),
```
Comment (frant-hartm) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java:232]: All the usages could use static import.

**Review by frant-hartm — CHANGES_REQUESTED** [review]
I meant to request changes :-)

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java (hunk 1 above)
Comment (TomaszGaweda) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java:59]: `wrap`? ;) Would make sense as there is `unwrap`

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import javax.annotation.Nullable;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Wraps Calcite's RexNodes, so that connector can be created outside hazelcast-sql module without problems with
+ * shaded Calcite dependency.
+ */
+public final class CalciteNode {
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java]: Sounds good - I didn't really have an idea to be honest ;)

**Comment by TomaszGaweda** [comment]
Thanks, all addressed @frant-hartm :)

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java (hunk 2 above)
Comment (TomaszGaweda) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java]: https://github.com/hazelcast/hazelcast/pull/23530/commits/8df0f28e672d92accf54d639013e431f15677d57

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java (hunk 1 above)
Comment (TomaszGaweda) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java]: https://github.com/hazelcast/hazelcast/pull/23530/commits/8df0f28e672d92accf54d639013e431f15677d57

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Wraps Calcite's RexNodes, so that connector can be created outside hazelcast-sql module without problems with
+ * shaded Calcite dependency.
+ */
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java]: https://github.com/hazelcast/hazelcast/pull/23530/commits/8df0f28e672d92accf54d639013e431f15677d57

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    /**
+     * Returns underlying Calcite {@linkplain RexNode}.
+     */
+    @Nullable
+    public <T> T unwrap(Class<T> clazz) {
+//        if (clazz.isInstance(node)) {
+//            return clazz.cast(node);
+//        }
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/CalciteNode.java]: https://github.com/hazelcast/hazelcast/pull/23530/commits/8df0f28e672d92accf54d639013e431f15677d57

**Review by frant-hartm — APPROVED** [review]

**Review by viliam-durina — APPROVED** [review]

**Pull request merged**